applications/LinearSolversApplication/custom_solvers/mkl_ilu.cpp

```diff
@@ -0,0 +1,403 @@
+//    |  /           |
+//    ' /   __| _` | __|  _ \   __|
+//    . \  |   (   | |   (   |\__ `
+//   _|\_\_|  \__,_|\__|\___/ ____/
+//                   Multi-Physics
+//
+//  License:         BSD License
+//                   Kratos default license: kratos/license.txt
+//
+//  Main authors:    Máté Kelemen
+//
+
+// External includes
+#include "mkl.h" // MKL_INT, dcsrilu0, dcsrilut
+#include "mkl_spblas.h" // mkl_sparse_trsv, mkl_sparse_destroy, mkl_sparse_d_create_csr
+
+// Project includes
+#include "custom_solvers/mkl_ilu.hpp" // MKLILUSmootherBase
+#include "spaces/ublas_space.h" // TUblasSparseSpace, TUblasDenseSpace
+
+// STL includes
+#include <vector> // std::vector
+#include <array> // std::array
+#include <algorithm> // std::fill
+#include <memory> // std::unique_ptr
+
+
+namespace Kratos {
+
+
+struct MKLCSRAdaptorDestructor {
+    void operator()(sparse_matrix_t* pAdaptor)
+    {
+        if (pAdaptor) {
+            mkl_sparse_destroy(*pAdaptor);
+            delete pAdaptor;
+        }
+        pAdaptor = nullptr;
```

Comment:
just to make the life easier to understand, i would move the `pAdaptor = nullptr` to the if block. just an opinion to avoid unnecessary operations.